cpp/include/proxy_sequences.h

```diff
@@ -118,7 +118,30 @@ class UpdateUploadObjectFromFile final : public Update {
   const alba::Checksum *_cs_o;
 };
 
-/* TODO upload object from in memory blob */
+class UpdateUploadObject final : public Update {
+public:
+  UpdateUploadObject(const std::string &name,
+                     const std::shared_ptr<std::vector<char>> data,
```

Comment:
The `shared_ptr` could go away as `mb.add_raw` below has a comment that it copies. Which leaves the vector - callers might not have the data in a vector, so a plain `const uint8_t` together with a `size_t` would be more generic.

Excursion: the CPP core guidelines / its support lib come with a `span` template:
https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#bounds1-dont-use-pointer-arithmetic-use-span-instead
https://github.com/Microsoft/GSL

Using the GSL for voldrv is eventually is on my wishlist, but that needs a more recent compiler than available to us at the moment (on Ubuntu 14.04).